modules/ding_adgangsplatformen/ding_adgangsplatformen.module

```diff
@@ -0,0 +1,421 @@
+  return array(
+    'auth' => array(
+      'required' => TRUE,
+      'install time setup' => FALSE,
+    ),
+  );
+}
+
+/**
+ * Implements hook_ajax_login_command().
+ */
+function ding_adgangsplatformen_ajax_login_command($path) {
+  // Javascript need to do the redirect is loaded in the init hook.
+  ctools_include('ajax');
+
+  // Get authentication url and set the path to be redirected back to.
+  $url = ding_adgangsplatformen_login(FALSE, $path);
+
+  // Store the ajax callback url that triggered the ajax auth redirect.
+  $parameters = $_GET;
+  unset($parameters['q']);
+  $_SESSION['oauth2ajax'] = $_GET['q'] . '?' . drupal_http_build_query($parameters);
+
+  return ctools_ajax_command_redirect($url);
+}
+
+/**
+ * Get the authentication configuration.
+ *
+ * @return array
+ *   With the default parameters if not overridden in the administration
+ *   settings form.
+ */
+function ding_adgangsplatformen_get_configuration() {
+  global $base_url;
+  return variable_get('ding_adgangsplatformen_settings', array(
+    'clientId' => '',
+    'clientSecret' => '',
+    'redirectUri' => $base_url . '/' . DING_ADGANGSPLATFORMEN_REDIRECT_URL,
+    'urlAuthorize' => 'https://login.bib.dk/oauth/authorize',
+    'urlAccessToken' => 'https://login.bib.dk/oauth/token/',
+    'urlResourceOwnerDetails' => 'https://login.bib.dk/userinfo/',
+    'urlLogout' => 'https://login.bib.dk/logout/',
+  ));
+}
+
+/**
+ * Login menu callback.
+ *
+ * Will start the authentication process and redirect the user to the
+ * authorization service.
+ *
+ * @param bool $redirect
+ *   Default TRUE, mainly used to get the auth url when used in ajax context
+ *   and an redirect is not desirable at this time.
+ * @param string $destination
+ *   Override the destination set in the current request. Mainly used in ajax
+ *   context.
+ *
+ * @return mixed
+ *    If $redirect is FALSE the authentication url or redirect headers.
+ *
+ * @throws Exception
+ *   If required libraries are not loaded.
+ */
+function ding_adgangsplatformen_login($redirect = TRUE, $destination = NULL) {
+  $provider = ding_adgangsplatformen_get_provider();
+  $authorization_url = $provider->getAuthorizationUrl();
+
+  // Get the state generated for you and store it to the session.
+  $_SESSION['oauth2state'] = $provider->getState();
+  $destination = empty($destination) ? $_REQUEST['destination'] : $destination;
+  ding_user_set_destination($destination);
+
+  // Check if an identity provider have been set in the request.
+  if (!empty($_REQUEST['idp'])) {
+    $authorization_url .= '&idp=' . $_REQUEST['idp'];
+  }
+
+  if (!$redirect) {
+    return $authorization_url;
+  }
+
+  // Redirect the user to the authorization URL. We don't use drupal_goto as it
+  // looks into the destination in the current URL and preforms some magic with
+  // it that we don't want.
+  header('Location:' . $authorization_url);
+  drupal_exit($authorization_url);
+}
+
+/**
+ * Logout the user from auth.
```

Comment:
Not sure that this would not just end up in two functions doing the same thing. The only place it's used as an menu callback is in the optional registration module (user can cancel the usage agreement page) and has a help when debugging to get logged out of SSO.